File: apps/tabtin-ios/Tabtin/Features/Conversation/ConversationTypography.swift
import SwiftUI
import UIKit

/// 会话阅读排版，对齐 Electron `chatDesignTokens`：
/// `CHAT_MESSAGE_TEXT_*` / `CHAT_STEP_TEXT` / `COMPOSER_TEXT_*`。
///
/// App chrome（导航、列表、设置）走 `TTFonts` 的 14pt `body`；
/// **本类型只服务对话阅读面**，避免长文与控件密度抢同一档。
///
/// 正典：`apps/tabtin-ios/docs/design-system.md` §3。
enum ConversationTypography {
    /// 消息正文 / Markdown：15pt（Electron `CHAT_MESSAGE_TEXT_BODY_BASE`）。
    static let bodySize: CGFloat = 15

    /// 一级标题（Electron `CHAT_MARKDOWN_HEADING_1` / `text-title`）。
    static let heading1Size: CGFloat = 20

    /// 二级标题（Electron `CHAT_MARKDOWN_HEADING_2`）。
    static let heading2Size: CGFloat = 18

    /// 步骤行固定行高 22pt（Electron `CHAT_STEP_TEXT`）。
    static let stepLineHeight: CGFloat = 22

    /// Markdown / 纯文本目标行高倍数（Electron `leading-[1.7]`）。
    static let bodyLineHeightMultiple: CGFloat = 1.7

    static let bodyUIFont = UIFont.systemFont(ofSize: bodySize, weight: .regular)

    static var bodyFont: Font { Font(bodyUIFont) }
    static var stepFont: Font { bodyFont }

    /// SwiftUI `Text` 在默认行高之上追加的间距。
    static var bodyLineSpacing: CGFloat {
        spacing(toLineHeight: bodySize * bodyLineHeightMultiple, font: bodyUIFont)
    }

    static var stepLineSpacing: CGFloat {
        spacing(toLineHeight: stepLineHeight, font: bodyUIFont)
    }

    /// MarkdownUI 段落相对行距 em（在 `MarkdownTheme` 中转为 `RelativeSize.em`）。
    /// 约等于在引擎默认行高上补到 ~1.7。
    static let markdownParagraphLineSpacingEm: CGFloat = 0.38

    static var composerFont: Font { bodyFont }
    static var composerLineSpacing: CGFloat { stepLineSpacing }

    /// Agent 身份牌 / Composer meta：走 UI 字号体系的 `meta`（13pt）。
    static var metaFont: Font { .tt.meta }

    private static func spacing(toLineHeight target: CGFloat, font: UIFont) -> CGFloat {
        max(0, target - font.lineHeight)
    }
}

File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/ConversationTypography.kt
package com.tabtin.mobile.features.conversation

import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import com.tabtin.mobile.ui.theme.TTFonts

/**
 * 会话阅读排版，对齐 Electron `chatDesignTokens`：
 * `CHAT_MESSAGE_TEXT_*` / `CHAT_STEP_TEXT` / `COMPOSER_TEXT_*`。
 *
 * App chrome（导航、列表、设置）走 [TTFonts] 的 14sp `body`；
 * **本对象只服务对话阅读面**，避免长文与控件密度抢同一档。
 *
 * 正典：`apps/tabtin-android/docs/design-system.md` §3。
 */
public object ConversationTypography {

    /** 消息正文 / Markdown：15sp（Electron `CHAT_MESSAGE_TEXT_BODY_BASE`）。 */
    public const val BODY_SIZE_SP: Float = 15f

    /** 一级标题（Electron `CHAT_MARKDOWN_HEADING_1` / `text-title`）。 */
    public const val HEADING_1_SIZE_SP: Float = 20f

    /** 二级标题（Electron `CHAT_MARKDOWN_HEADING_2`）。 */
    public const val HEADING_2_SIZE_SP: Float = 18f

    /** 二级标题行高（Electron `leading-[26px]`）。 */
    public const val HEADING_2_LINE_HEIGHT_SP: Float = 26f

    /** 步骤行固定行高 22sp（Electron `CHAT_STEP_TEXT`）。 */
    public const val STEP_LINE_HEIGHT_SP: Float = 22f

    /** Markdown / 纯文本目标行高倍数（Electron `leading-[1.7]`）。 */
    public const val BODY_LINE_HEIGHT_MULTIPLE: Float = 1.7f

    /** Markdown 段落相对行距 em（在 Markdown 主题中换算）。 */
    public const val MARKDOWN_PARAGRAPH_LINE_SPACING_EM: Float = 0.38f

    private val bodyLineHeightSp: Float = BODY_SIZE_SP * BODY_LINE_HEIGHT_MULTIPLE

    /** 消息正文 / Markdown 基础样式。 */
    public val body: TextStyle = TextStyle(
        fontSize = BODY_SIZE_SP.sp,
        lineHeight = bodyLineHeightSp.sp,
        fontWeight = FontWeight.Normal,
    )

    public val bodySemibold: TextStyle = body.copy(fontWeight = FontWeight.SemiBold)

    /** 思考 / 工具步骤行。 */
    public val step: TextStyle = TextStyle(
        fontSize = BODY_SIZE_SP.sp,
        lineHeight = STEP_LINE_HEIGHT_SP.sp,
        fontWeight = FontWeight.Normal,
    )

    public val stepSemibold: TextStyle = step.copy(fontWeight = FontWeight.SemiBold)

    /** Composer 主输入（同步骤行 metrics）。 */
    public val composer: TextStyle = step

    /** Agent 身份牌 / Composer meta：走 UI 字号体系的 `meta`（13sp）。 */
    public val meta: TextStyle = TTFonts.meta
}
